### PR TITLE
Rename `BasicAer` to `BasicProvider` and refactor interface to follow `BackendV2`

### LINKED ISSUE #4443: Rename BasicAer as a test provider
BasicAer is not needed anymore now that the `quantum_info` module can be used to simulate circuits. `quantum_info` has a more flexible workflow as you are not tied to the job/result interface and is nice for small circuit calculations. It is also faster as it bypasses serialization.

`Operator(qc)` (unitary) or `zero.evolve(qc)` (statevector) or `zero.evolve(qc).probabilities_dict()` (counts)

The only thing that BasicAer supports in addition to `quantum_info` is conditional operations I think, but I think that can be added too?

### LINKED ISSUE #10690: Use `BackendV2` for basic aer
### What should we add?

Right now the backends in `qiskit.providers.basicaer` are all built on the legacy `BackendV1` model. We should update it to be based on the more modern `BackendV2` instead. The target will be very simple for these backends, it's basically just:

```python
MAX_QUBITS_MEMORY = int(log2(local_hardware_info()["memory"] * (1024**3) / 16))
target = Target(num_qubits=min(24, MAX_QUBITS_MEMORY)
lam = Parameter("lambda")
target.add_instruction(U1Gate(lam))
phi = Parameter("phi")
target.add_instruction(U2Gate(phi, lam))
theta = Parameter("theta")
target.add_instruction(U3Gate(theta, phi, lam))
target.add_instruction(RZGate(phi))
target.add_instruction(SXGate())
target.add_instruction(XGate())
target.add_instruction(CXGate())
target.add_instruction(IGate())
target.add_instruction(UnitaryGate, name="unitary")
```

The trickier piece here is more about maintaining backwards compatibility for all the pieces that are part of the `BackendV1` interface and not in the `BackendV2` interface to give users a chance to migrate.

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
This PR introduces the following changes:
- Renames the `basicaer` module to `basic_provider` to avoid confusions with `qiskit-aer`.
- Removes `UnitarySimulatorPy` and `StatevectorSimulatorPy` from `basic_provider` in favor of the `quantum_info` equivalents (`Statevector` and `Operator`).
- Renames the `QasmSimulatorPy` class to `BasicSimulator` to avoid misinterpretations of it being a "qasm-based" simulator.
- Refactors the `BasicSimulator` to follow the `BackendV2` interface, a step forward in the deprecation of `BackendV1` (it keeps some legacy methods/properties to interface with `assemble` and other V1-centric functionality) --> fixes #10690.

Fixes https://github.com/Qiskit/qiskit/issues/4443 and #10690.

### Details and comments
The discussion in #4443 starts off considering renaming `BasicAer` to `TestProvider`, but I believe that `Test` is an overloaded term (at least, more than `Basic`) and it doesn't really represent the purpose of this module, which is offering a lightweight simulator that is shipped with qiskit for quick prototyping, not performance or efficiency (also testing, but that is not its only use). 

In addition to this, the test file names would look particularly ugly if it were called `TestProvider`, with paths like `test.providers.test_provider.test_test_provider.TestTestProvider` (as opposed to `test.providers.basic_provider.test_basic_provider.TestBasicProvider`).

### Mini migration guide:

```
Old                    |        New
---------------------------------------------------
basicaer               |  basic_provider
BasicAerProvider       |  BasicProvider 
BasicAerJob            |  BasicProviderJob
QasmSimulatorPy        |  BasicSimulator
UnitarySimulatorPy     |  use quantum_info instead
StatevectorSimulatorPy |  use quantum_info instead

```

* The corresponding deprecation PR will follow in the 0.46 branch. -> #11583

### CODE REVIEW

**Comment by MozammilQ** [comment]
@ElePT , does this PR targets issue #10690 ?

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]:
```suggestion
Where the input is a ``QuantumCircuit`` object and the output is a ``BasicProviderJob`` object,
```

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Just a suggestion for the consistent one-line summary, feel free to ignore or adapt 🙂 
```suggestion
    """Python implementation of a basic (slow, non-efficient) quantum simulator. 
    
    This implementation was originally based on the :class:`.BackendV1` interface, and later
    migrated to :class:`.BackendV2`. However, certain legacy methods and properties from 
    :class:`.BackendV1` have been kept for backwards compatibility purposes, these are:
```

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Since the typehints are already inline, can we skip them in the doc string?

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Are these V1-specific properties to remain or are they going away? I.e. is this class both BackendV1 and V2 or only the latter? 🙂

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Does this else-clause only ever affect the `required` operations? If yes, would it be cleaner to separate the for loop? If no, shouldn't the number of qubits depend on the gate?

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Should the gate for `"unitary"` be `UnitaryGate`?

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Small comment, but since the type is documented in the signature I think it doesn't need to be here
```suggestion
            The configuration for the backend.
```

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py:210]: This is just a question: does `max_shots=0` imply a statevector simulation? Or why is it set to 0?

File: qiskit/providers/basic_provider/basic_provider.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_provider.py]: Test or basic? 🙂 
```suggestion
"""Provider for basic simulator backends, formerly known as `BasicAer`."""
```

File: qiskit/providers/basic_provider/basic_provider.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_provider.py:38]: Is this still a V1 provider or also a V2 provider? If it is the latter, would it be an option to inherit from both `ProviderV1` and `ProviderV2`?

File: qiskit/providers/basic_provider/basic_provider_tools.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_provider_tools.py:29]: You didn't touch this in this PR, but it would be much nicer to implement this via something like
```
from qiskit.circuit.library import get_standard_gate_name_mapping

gates = get_standard_gate_name_mapping()
if gate not in gates.keys():
    raise ValueError ....
return gates[gate](*params).to_matrix()
```
but we could also do that in a separate small PR 🙂

File: test/python/circuit/test_initializer.py
Comment (Cryoris) [inline test/python/circuit/test_initializer.py]: While we're at it, should we remove the usage of `execute` in favor of `Backend.run`?

File: test/python/qobj/test_qobj_identifiers.py
Comment (Cryoris) [inline test/python/qobj/test_qobj_identifiers.py]: I think  I'm missing something obvious, but how does this test relate to the `BasicSimulator`?

File: test/randomized/test_synthesis.py
Comment (Cryoris) [inline test/randomized/test_synthesis.py:66]: What did this `seed` do? 🙂

**Review by Cryoris — COMMENTED** [review]
The overall approach LGTM, I left some comments below. 

One feature that we're losing by using `Statevector` over `StatevectorSimulatorPy` is that we cannot use statevector simulation of circuits with measurements anymore. Or is this included in the `BasicSimulator` using `shots=0`? In any case, we should make sure to point this out in the reno 🙂

File: qiskit/providers/basic_provider/basic_provider.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider.py]: true, that's an oversight, I had originally called it "test simulator".

File: qiskit/providers/basic_provider/basic_provider.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider.py:38]: I haven't found any `ProviderV2` class implementation. I was also confused by this, but the `IBMProvider` in `qiskit-ibm-provider` also inherits from `ProviderV1`. My understanding here is that the provider versions have nothing to do with the backend versions, and that instead of increasing provider versions what we have are different kinds of providers (like the runtime service).

File: qiskit/providers/basic_provider/basic_provider_tools.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider_tools.py:29]: I could look into it. Thanks!

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_simulator.py]: This is a `BackendV2` with some V1-specific properties that will go away within the next year. Ideally we should not need them, but the two interfaces are too entangled to change now.

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_simulator.py]: this section of the code is definitely improvable, I'll try to make it cleaner one I figure out why the windows tests are getting stuck.

File: test/python/circuit/test_initializer.py
Comment (ElePT) [reply test/python/circuit/test_initializer.py]: I believe this is taking place in a separate PR.

File: test/randomized/test_synthesis.py
Comment (ElePT) [reply test/randomized/test_synthesis.py:66]: It was used as a seed to the unitary simulator, which is now no longer used (I don't think `Operator` takes a seed).

File: qiskit/providers/basic_provider/basic_provider_tools.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider_tools.py:29]: Given that `get_standard_gate_name_mapping()` gives you already initialized gates, and this method initializes them, I'd rather keep it for a potential follow-up.

File: test/python/qobj/test_qobj_identifiers.py
Comment (ElePT) [reply test/python/qobj/test_qobj_identifiers.py]: `QasmSimulatorPy` has been renamed to `BasicSimulator` :)

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_simulator.py:210]: This is a good question, as far as I can see, this attribute is only used by the `assembler` https://github.com/Qiskit/qiskit/blob/fc567f0883239d416bfdea55a1e5ad2124ab4e61/qiskit/compiler/assembler.py#L304 This line was directly taken from the previous implementation, and seems to be unrelated to `self._shots`, which are applied manually during the run method. So my take here is, this is done to save some work during `assemble`  I would not worry too much about it given that both the BackendConfiguration and assemble should go away soon.

File: test/python/qobj/test_qobj_identifiers.py
Comment (Cryoris) [reply test/python/qobj/test_qobj_identifiers.py]: This test doesn't seem to be covered in `test_qobj.py` so maybe we can just rename this to `test_qobj_identifiers` or something until Qobj is gone

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Do you want to use Sphinx links here? (I'm not putting an inline suggestion as that would probably break the 100 chars line length 😄)

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: I think "slow" is relative to what you want to simulate and it would raise the question why we provide this -- maybe just leaving "basic" as adjective is better?
```suggestion
    """Python implementation of a basic quantum simulator.
```

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py]: Oftentimes we remove the first line if it doesn't state something special. It also looks good in the docs since after the classname the args are directly listed 🙂 Just an optional suggestion!
```suggestion
        """
```

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [reply qiskit/providers/basic_provider/basic_simulator.py:210]: In this case, should we leave an inline comment saying that this is for `assemble`?

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (Cryoris) [inline qiskit/providers/basic_provider/basic_simulator.py:406]: Just a question, nothing to change: I'm a bit confused that this is explicitly handled here but we don't maintain `Initialize` instructions during transpilation

File: qiskit/providers/fake_provider/fake_backend.py
Comment (Cryoris) [inline qiskit/providers/fake_provider/fake_backend.py]: Since this message gives information about what simulator is used, would it be clearer to state that the `BasicSimulator` is used?

File: qiskit/transpiler/passes/layout/full_ancilla_allocation.py
Comment (Cryoris) [inline qiskit/transpiler/passes/layout/full_ancilla_allocation.py]: Not required but explicitly checking for `None` is a bit safer (and would've prevented some bugs in the past 🙂)
```suggestion
        if self.target is not None and self.target.num_qubits is not None:
```

File: qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py
Comment (Cryoris) [inline qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py]: For consistency (`is not None` is used above) and safety 🙂 
```suggestion
        if self._target is not None and self._target.num_qubits is not None:
```

File: qiskit/transpiler/target.py
Comment (Cryoris) [inline qiskit/transpiler/target.py]: Should this check for `None`? As it is I think it's the correct behavior for both `None` and `0` being explicit might be clearer

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (Cryoris) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Is this going to be introduced in this PR or in the 0.46 PR?

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (Cryoris) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Uppercase `Target` (w/o the link) would also be good I think
```suggestion
    creation of :class:`.Target`-based simulators with a flexible number of qubits.
```

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (Cryoris) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: This sounds like the `quantum_info` module is new, maybe rephrase like this?
```suggestion
    replaced by the :mod:`qiskit.quantum_info` module and the new 
    :mod:`qiskit.providers.basic_provider` module. The migration from using
```

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (Cryoris) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Nice!

File: test/python/transpiler/test_1q.py
Comment (Cryoris) [inline test/python/transpiler/test_1q.py]: Removing the `test_` prefix will cause this test to be skipped, right?

File: test/python/transpiler/test_passmanager_config.py
Comment (Cryoris) [inline test/python/transpiler/test_passmanager_config.py]: See above 🙂

**Review by Cryoris — COMMENTED** [review]
Thanks for addressing all comments, this mostly looks good. I left some very small comments below 🙂 Also we should still update the `docs/apidoc/providers_basicaer.rst`  description

File: test/python/transpiler/test_1q.py
Comment (ElePT) [reply test/python/transpiler/test_1q.py]: oversight from a search and replace

File: qiskit/providers/basic_provider/basic_simulator.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_simulator.py]: I get your point about slow, but think I can leave the non-efficient part.

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (ElePT) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]:
```suggestion
    creation of :class:`.Target` based simulators with a flexible number of qubits.
```

**Review by Cryoris — DISMISSED** [review]
LGTM, thanks for all the work! Marking as  `on hold` until we merged the deprecation PR to 0.46 🙂

File: examples/python/using_qiskit_terra_level_0.py
Comment (ElePT) [inline examples/python/using_qiskit_terra_level_0.py]:
```suggestion
and runs them using a simulator in the BasicProvider (local Qiskit provider).
```

File: qiskit/providers/backend_compat.py
Comment (mtreinish) [inline qiskit/providers/backend_compat.py]: Was there a case where `BackendConfiguration` could not have `n_qubits`, IIRC it's a required attribute that's built into the model class.

File: qiskit/providers/basic_provider/basic_provider.py
Comment (mtreinish) [inline qiskit/providers/basic_provider/basic_provider.py]: Are there any backends that need this. I realize this is what we merged on 0.46, but isn't the `backends` list just `[BasicSimulator()]` now with no aliases or deprecations? I feel like this logic can be reduced quite a bit and just reduced to:

```suggestion
```

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (mtreinish) [reply releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Yeah, this feature note should be in 0.46.0 which is the first release with the basic provider.

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, it's mostly identical to the additions in the #11583 just a couple small questions inline. But otherwise this is ready to go.

File: qiskit/providers/basic_provider/basic_provider.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider.py:81]: I removed the alias resolution  from both `backends` and `get_backend` (which ended up being just `super().get_backend`).

File: qiskit/providers/backend_compat.py
Comment (ElePT) [reply qiskit/providers/backend_compat.py]: I added this at the beginning when the `BasicSimulator` was a bit of a hybrid between V1 and V2, and forgot to change it back. No longer used or needed.

File: qiskit/providers/basic_provider/basic_provider.py
Comment (ElePT) [reply qiskit/providers/basic_provider/basic_provider.py]: Removed alias shenanigans in [85addc3](https://github.com/Qiskit/qiskit/pull/11422/commits/85addc33bcde64bd858874611eeecd5a059dce2d) (I had to add a bit of logic because `get_backend()` relies on `backends()`, but I think it's still better now).

File: qiskit/providers/basic_provider/basic_provider.py
Comment (mtreinish) [inline qiskit/providers/basic_provider/basic_provider.py]:
```suggestion
from qiskit.providers.providerutils import filter_backends
```

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (mtreinish) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Did you want to mention `qiskit.BasicAer` here being removed too?

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (mtreinish) [inline releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]:
```suggestion
        ``qiskit.providers.basicaer` |  :mod:`.basic_provider`
        ``BasicAerProvider``         |  :class:`.BasicProvider`
        ``BasicAerJob``              |  :class:`.BasicProviderJob`
        ``QasmSimulatorPy``          |  :class:`.BasicSimulator`
        ``UnitarySimulatorPy``       |  use :class:`~.quantum_info.Operator`
        ``StatevectorSimulatorPy``   |  use :class:`~.quantum_info.Statevector`
```

**Review by mtreinish — DISMISSED** [review]
This LGTM, just some quick questions in the release note. But we can save this for a followup post rc1 if you'd prefer as we'll be going through the release notes for the final release.

File: releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml
Comment (ElePT) [reply releasenotes/notes/refactor-basicaer-to-basicprovider-e27aff9c8e81d26e.yaml]: Done!

**Review by kevinhartman — APPROVED** [review]

**Pull request merged**
